lib/chai/core/assertions.js

```diff
@@ -2448,135 +2448,138 @@ module.exports = function (chai, _) {
   Assertion.addMethod('key', assertKeys);
 
   /**
-   * ### .throw([errorLike], [errMsgMatcher], [msg])
+   * ### .throw
+   * ### .throw([errLike[, errMsgMatcher[, msg]]])
+   * ### .throw([errMsgMatcher[, msg]])
    *
-   * When no arguments are provided, `.throw` invokes the target function and
-   * asserts that an error is thrown.
+   * When used as a property assertion or when invoked without any arguments,
+   * `.throw` invokes the target function and asserts that it throws.
    *
-   *     var badFn = function () { throw new TypeError('Illegal salmon!'); };
+   *     var badFn = function () { throw TypeError('Illegal salmon!'); };
    *
+   *     expect(badFn).to.throw;
```

Comment:
This is an interesting addition - but it is also one we tried very much in vain to do back in Chai@2 days. The main issue here is that property assertions that assert and also return a function to assert messes a lot of stuff up. There were some terrible nasty bugs that came from this.

Aside from my concerns of implementation complexity, I also don't think this is actually internally consistent with the property-vs-assertion rule we have: `throw` takes arguments, and so having it as a property sometimes, and other times not, is something that does not have existing precedent.

Further to your point:

> I'm opposed to jumping to the opposite extreme of removing all property-based assertions from Chai; despite these types of problems, I think property-based assertions are a critical part of Chai's identity and appeal for many users.

I'm of the opposite mind set. For one, we see a lot of issues around implicitly passing tests because people often mistake function-assertion for a property-assertion, and get many bug reports about this. For another it makes the learning cliff a lot harder, as we're probably the _only_ library in the users tool-belt that has this behaviour. If it were a common idiom I'd be more behind it, but it is not. Another reason I'd like to see this pattern disappear is that it makes plugin development more complex, both from the API complexity but also from a philosophical perspective where not every plugin strictly adheres to the property-vs-method rubric, which of course in turn decreases usability because Chai's own rules aren't consistent.

Currently, with chai@5 there exists no way to handle property assertions, although I plan to add them in but via an optional interface, and that is forced based on something we can introspect in code (e.g. property assertions are made from any unary assertions - i.e one that only takes only an `actual` and not any additional arguments).

Reply:
My two main arguments against making property assertions "opt-in" in chai@5 are:

1. The history & popularity of Chai undermine the effectiveness of dropping property-based assertions this late in the game. Chai has had 7 years worth of blog posts and tutorials published on the Internet with assertions like `expect(blah).to.be.true`, so if a user unknowingly follows one of those outdated tutorials in chai@5, they now have an assertion that looks legit and has increased legitimacy from being a style that's demonstrated across many tutorials but now in chai@5 passes no matter what.
2. This type of bug report dropped drastically in chai@4 with the introduction of proxy protection. There's now only a couple of remaining assertions in Chai core that could be reasonably written such that they look legit but pass no matter what. `expect(fn).to.throw;` and `expect('meow').to.be.a.string;`. I'm not familiar with the implementation complexity and bugs that popped up when trying to fix this in chai@2; I guess I'd have to experience that pain myself to understand!

Good points about increased complexity for plugin authors, and consistency issues across both the Chai plugin ecosystem as well as JavaScript as a whole. I'm convinced that dropping property-based assertions would've been the right decision at the start of the project, but I'm not so sure it is now given Chai's history & popularity. That's a tough one.